x/concentrated-liquidity/pool_hooks.go

```diff
@@ -0,0 +1,95 @@
+package concentrated_liquidity
+
+import (
+	"github.com/cosmos/cosmos-sdk/store/prefix"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	types "github.com/osmosis-labs/osmosis/v20/x/concentrated-liquidity/types"
+)
+
+// nolint: unused
+// callPoolActionListener processes and dispatches the passed in message to the contract corresponding to the hook
+// defined by the given pool ID and action prefix (e.g. pool Id: 1, action prefix: "beforeSwap").
+//
+// This function returns an error if the contract address in state is invalid (should be impossible) or if the contract execution fails.
+//
+// Since it is possible for this function to be triggered in begin block code, we need to directly meter its execution and set a limit.
+// If no contract is linked to the hook, this function is a no-op.
+func (k Keeper) callPoolActionListener(ctx sdk.Context, msgBz []byte, poolId uint64, actionPrefix string) (err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			err = types.ContractHookOutOfGasError{GasLimit: types.ContractHookGasLimit}
+		}
+	}()
+
+	cosmwasmAddress := k.GetPoolHookContract(ctx, poolId, actionPrefix)
+	if cosmwasmAddress != "" {
+		cwAddr, err := sdk.AccAddressFromBech32(cosmwasmAddress)
+		if err != nil {
+			return err
+		}
+
+		em := sdk.NewEventManager()
+
+		// Since it is possible for this hook to be triggered in begin block code, we need to
+		// directly meter its execution and set a limit. See comments on `ContractHookGasLimit`
+		// for details on how the specific limit was chosen.
+		//
+		// We ensure this limit only applies to this call by creating a child context with a gas
+		// limit and then metering the gas used in parent context once the operation is completed.
+		childCtx := ctx.WithGasMeter(sdk.NewGasMeter(types.ContractHookGasLimit))
+		_, err = k.contractKeeper.Sudo(childCtx.WithEventManager(em), cwAddr, msgBz)
+		if err != nil {
+			return err
+		}
+
+		// Consume gas used for calling contract to the parent ctx
+		ctx.GasMeter().ConsumeGas(childCtx.GasMeter().GasConsumed(), "Track CL action contract call gas")
+	}
+
+	return nil
+}
+
+// --- Store helpers ---
+
+// GetPoolHookPrefixStore returns the substore for a specific pool ID where hook-related data is stored.
+func (k Keeper) GetPoolHookPrefixStore(ctx sdk.Context, poolID uint64) sdk.KVStore {
```

Comment:
Do we need to have this exported?

Reply:
Good point, wasn’t necessary. Also unexported `GetPoolHookContract` 46aa7f26684c44b455130981a18944005b331e6c